File: gcode/src/main/java/org/qw3rtrun/p3d/g/decoder/FirmwareReportDecoder.java
package org.qw3rtrun.p3d.g.decoder;

import org.apache.commons.lang3.StringUtils;
import org.qw3rtrun.p3d.g.event.FirmwareInfoReportEvent;
import org.qw3rtrun.p3d.g.event.FirmwareReport;

import java.util.HashMap;
import java.util.Optional;
import java.util.regex.Pattern;

public class FirmwareReportDecoder implements GEventDecoder<FirmwareInfoReportEvent> {

    private static String FW_PREFIX = "FIRMWARE_NAME";

    private static String SEPARATOR = ":";

    private static Pattern PATTERN = Pattern.compile("((?<field>[A-Z_]+)" + SEPARATOR + ")+");

    @Override
    public Optional<FirmwareInfoReportEvent> decode(String line) {
        if (!StringUtils.startsWithIgnoreCase(line, FW_PREFIX + SEPARATOR)) {
            return Optional.empty();
        }
        var matcher = PATTERN.matcher(line);
        var map = new HashMap<String, String>();
        if (!matcher.find()) {
            return Optional.empty();
        }
        String key = matcher.group();
        int keyEnd = matcher.end();
        while (matcher.find()) {
            map.put(key.substring(0, key.length() - 1), line.substring(keyEnd, matcher.start()).trim());
            key = matcher.group();
            keyEnd = matcher.end();
        }
        map.put(key.substring(0, key.length() - 1), line.substring(keyEnd, line.length()));
        return Optional.of(new FirmwareReport(map));
    }

}

File: gcode/src/main/kotlin/org/qw3rtrun/p3d/g/marlin/decoder/FirmwareReportDecoder.kt
package org.qw3rtrun.p3d.g.marlin.decoder

import org.qw3rtrun.p3d.core.msg.FirmwareInfoReportEvent
import org.qw3rtrun.p3d.core.msg.FirmwareReport
import java.util.Optional
import java.util.regex.Pattern

class FirmwareReportDecoder : GEventDecoder<FirmwareInfoReportEvent> {

    override fun decode(line: String): Optional<FirmwareInfoReportEvent> {
        if (!line.startsWith(FW_PREFIX + SEPARATOR, ignoreCase = true)) {
            return Optional.empty()
        }
        val matcher = PATTERN.matcher(line)
        val map = mutableMapOf<String, String>()
        if (!matcher.find()) {
            return Optional.empty()
        }
        var key = matcher.group()
        var keyEnd = matcher.end()
        while (matcher.find()) {
            map[key.substring(0, key.length - 1)] = line.substring(keyEnd, matcher.start()).trim()
            key = matcher.group()
            keyEnd = matcher.end()
        }
        map[key.substring(0, key.length - 1)] = line.substring(keyEnd, line.length)
        return Optional.of(FirmwareReport(map))
    }

    companion object {
        private const val FW_PREFIX = "FIRMWARE_NAME"
        private const val SEPARATOR = ":"
        private val PATTERN = Pattern.compile("((?<field>[A-Z_]+)$SEPARATOR)+")
    }
}
